core/platform/src/error_code.rs

```diff
@@ -1,25 +1,36 @@
-/// An error code returned by the kernel. Tock's system calls return errors as a
-/// negative `isize`. This wraps the isize, and is useful for adding type safety
-/// to APIs.
-
+/// A system call error code. This can either be an error code returned by the
+/// kernel or BADRVAL, which indicates the kernel returned the wrong type of
+/// response to a system call.
+// ErrorCode is not an enum so that conversion from the kernel's return value (a
+// `usize` in a register) is free.
```

Comment:
It's true that the "conversion" from a usize is essentially free here. However, wouldn't you at some point want a checked enumeration of all of the different variants (e.g. a Rust enum) still? I'm not very familiar with the architecture, so that might of course be planned further down the line.

In any case, I'd appreciate naming it accordingly (something like `UncheckedErrorCode`) to indicate that this is not yet verified to be a valid (defined) `ErrorCode` variant. The same would go for `ReturnType`.

Reply:
I see that the raw values make sense, for example in the comparison of the `raw_yield` return type in `fn yield_no_wait(`. Still, will there at some point be a checked enum?

Reply:
I don't see the use case for a checked enumeration for `ReturnType`. Every system call should only have one success type and one failure type, a driver's return type handling will have at most 3 cases:

1. The expected success type
2. The expected failure type
3. A default case that returns something like BADRVAL

Checking that the `ReturnType` is a known variant in `libtock_platform` seems like a waste of instruction space, as the code that interprets the `ReturnType` doesn't care about the difference between an unknown variant and a variant it didn't expect.

As for `ErrorCode`, I was operating under the assumption that we may add additional error codes to the [error codes table](https://github.com/tock/tock/blob/tock-2.0-dev/doc/reference/trd-syscalls.md#32-error-codes) in the future. If that list is final, then I think there is a benefit to making `ErrorCode` a checked enum.

Do we expect the Tock 2.0 error code list to be permanent (i.e. that we will not add another kernel error code after Tock 2.0 and before Tock 2.0)?

Reply:
I'm not aware of any error code to be added in the immediate future, but we shouldn't promise it won't be expanded.

The reasoning behind this architecture is perfectly valid though. For all interactions with the system call ABI itself, the raw (unchecked) `ReturnType` should be absolutely sufficient.

I'm more concerned about `ErrorCode` as I imagine that would be used in application code further down the line. When using it in application code and branching depending on the precise error returned (for instance in a `match` expression), Rust can force you to cover all cases, which is especially helpful to discover code segments requiring updates if new error code variants would be introduced. However, given the large number of error codes already, one would most certainly use a fallback match arm, so that benefit does not play out. In the end your decision towards minimizing code size seems reasonable, thanks for explaining it!